src/axl_async_nnfdm.c

```diff
@@ -0,0 +1,224 @@
+#include <unistd.h>           /* For: sleep */
+#include "axl_internal.h"
+#include "datamovement.pb-c.h"
+#include "kvtree.h"
+#include "nnfdm.h"
+
+static GoInt32 nnfdm_connection_ref = 0;
+static int nnfdm_initialized = 0;
+
+/* TODO: Move to appropriate header file for key/value keys */
+#define AXL_KEY_FILE_SESSION_UID "NNFDM_Session_ID"
+
+
+void axl_async_init_nnfdm() {
+    if (!nnfdm_initialized) {
+        nnfdm_initialized++;
+        char *socketAddr = "/var/run/nnf-dm.sock";
+
+        struct OpenConnection_return conn;
+        conn = OpenConnection(socketAddr);
+        if (conn.r1 != 0) {
+            axl_abort(-1,
+                "NNFDM OpenConnection(%s) failed with error %d @ %s:%d",
+                socketAddr, conn.r1, __FILE__, __LINE__
+            );
+        }
+        nnfdm_connection_ref = conn.r0;
+    }
+}
+
+void axl_async_finalize_nnfdm() {
+    if (nnfdm_initialized) {
```

Comment:
Another option would be to decrement the reference counter here, and finalize on 0.  I'm not sure if we should actually finalize on the first or last call.